### PR TITLE
Add Codegen Support

### LINKED ISSUE #7: Support new dnd methods
Documented at: https://api.slack.com/methods#dnd

### PR DESCRIPTION
**Pull request by mthjones**
With Rust 1.15 now released, this can now build on stable. I figured it's time to open it up for review

An overview of what's been done:
* Literally everything has been replaced
* All methods are auto-generated from API schema files
* All types (request params, response objects, errors) are auto-generated from API schema files
* Now uses serde for (de)serialization
* Now uses reqwest as the default client (hyper dropped openssl support, which is necessary, and reqwest uses rust-native-tls)
* No more useless tests

Highly recommend checking the branch out and running `cargo doc` to view the new documentation and get an idea of what's changed.

Some thoughts:
* `files.upload` - we still don't have a great way to support this. I think we should just elide it from the generated output and we can feature-flag it later once we actually implement support
* Messages are still a bit rough around the edges. Should just require updating the schemas and updating the submodule here.
* Do we need dynamic codegen? Or can we just codegen, copy it to the src folder and release that? Dynamic codegen helps if we want to, say, add features so users can pick which modules they want, but is otherwise a hindrance to grokking the code.
* Basically every response field is an `Option<T>` now. I don't trust Slack enough anymore to not just remove fields on a whim.

Closes #7, #8, #9, #10, #31

### CODE REVIEW

**Comment by BenTheElder** [comment]
I agree with all of this. This looks pretty excellent at a glance, I'll have to look at this in detail tomorrow.

For the codegen: I've only toyed with this a little. Is it usual in rust to store codegen output?
I've done something similar for smaller projects in other languages, but I'm not aware of the norms for rust in this. I think ultimately that it would be very useful to have codegen stored with the project for grokking as you mentioned so I'd be all for that. Perhaps we should follow whatever the rust community norms are though.

This looks like really excellent work. Thank you for taking this on. I will take a closer look sometime tomorrow afternoon to catch up with everything.

**Comment by BenTheElder** [comment]
Other than the awesome work of having the api generated from a proper schema, this point Especially I think is an important change we should move to:
> Basically every response field is an Option<T> now. I don't trust Slack enough anymore to not just remove fields on a whim.

:+1:

**Comment by mthjones** [comment]
> Is it usual in rust to store codegen output?
Mostly, I just modeled it off what [`rusoto`](https://github.com/rusoto/rusoto) does. Since they are generating code off the AWS schemas, I figured basing this work off that made sense. However, they also require opting into certain features and only building those -- something that makes sense for an AWS library (generally not using API) but may not for a Slack one.

> an important change we should move to
:) It's not great to work with, but without Slack releasing a properly versioned API, it's the best we can do. Alternatively, if Slack were to pick up the schemas, they can update which fields are required themselves. The support is there, just requires adding the field to the `required: []` array in the schema.

Thanks for taking a look! Looking forward to the feedback. Admittedly some of the code is pretty hacky in `generator.rs`. I haven't done a proper refactoring pass yet.

**Comment by BenTheElder** [comment]
I think we should definitely include the codegen output then.

Have you raised an issue with slack about potentially supporting the schemas? I'm sure some other projects would be interested in them as well.

File: codegen/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[package]
```
Comment (BenTheElder) [inline codegen/Cargo.toml]: does crates.io now allow publishing packages with sub crates like this?

Comment (BenTheElder) [inline codegen/Cargo.toml]: Never-mind. It seems like at one point you couldn't do this, but it appears to be somewhat common now.

**Comment by BenTheElder** [comment]
I haven't quite had time to fully absorb the codegen yet, but so far this looks excellent :+1:

**Comment by mthjones** [comment]
`serde@0.9.6` just released with internally tagged enum support! Going to look at integrating that into this to remove the hacky custom deserialization of enums I had to do.

**Comment by mthjones** [comment]
I'm fairly comfortable with where this sits now. Once [this upstream PR](https://github.com/whatisinternet/inflector/pull/46) is merged and released, I'll merge this in. There are a couple minor issues outstanding that I can file as issues after merging.

**Comment by BenTheElder** [comment]
Awesome! Looking forward to getting this in and seeing the whole project
overhauled.

**Comment by mthjones** [comment]
I've worked around the upstream issue for now. Once it gets merged and released upstream, the code generator can be updated to use it without it needing to be a breaking change here.

Time to merge!

**Comment by BenTheElder** [comment]
Whoot!
